ignition-server/controllers/machineconfigserver_ignitionprovider.go

```diff
@@ -0,0 +1,338 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	k8sutilspointer "k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ IgnitionProvider = (*MCSIgnitionProvider)(nil)
+
+// MCSIgnitionProvider is an IgnitionProvider that uses
+// MachineConfigServer pods to build ignition payload contents.
+type MCSIgnitionProvider struct {
+	Client          client.Client
+	ReleaseProvider releaseinfo.Provider
+	Namespace       string
+}
+
+func (p *MCSIgnitionProvider) GetPayload(ctx context.Context, tokenSecret *corev1.Secret) (payload []byte, err error) {
+	// TODO(alberto): If the MCS supports binding address
+	// https://github.com/openshift/machine-config-operator/pull/2630/files
+	// we could bind it to localhost and get the payload by execing
+	// https://zhimin-wen.medium.com/programing-exec-into-a-pod-5f2a70bd93bb
+	// Otherwise with current approach the mcs pod is temporary exposed in the pod network every time a payload is generated.
+	releaseImage := string(tokenSecret.Data[TokenSecretReleaseKey])
+	img, err := p.ReleaseProvider.Lookup(ctx, releaseImage)
+	if err != nil {
+		return nil, fmt.Errorf("failed to look up release image metadata: %w", err)
+	}
+
+	mcsResourceName := tokenSecret.GetName()
+	mcsServiceAccount := MachineConfigServerServiceAccount(p.Namespace, mcsResourceName)
+	mcsRoleBinding := MachineConfigServerRoleBinding(p.Namespace, mcsServiceAccount)
+	mcsPod := mcsPod(p.Namespace, mcsResourceName, img, mcsServiceAccount)
+	// Launch the pod and ensure we clean up regardless of outcome
+	if err := p.Client.Create(ctx, mcsPod); err != nil {
+		return nil, fmt.Errorf("failed to create machine config server pod: %w", err)
+	}
+
+	if err := p.Client.Create(ctx, mcsServiceAccount); err != nil {
+		return nil, fmt.Errorf("failed to create machine config server Service Account: %w", err)
+	}
+
+	// Launch the pod and ensure we clean up regardless of outcome
+	if err := p.Client.Create(ctx, mcsRoleBinding); err != nil {
+		return nil, fmt.Errorf("failed to create machine config server Role Binding: %w", err)
+	}
+	defer func() {
+		var deleteErrors []error
+		if err := p.Client.Delete(ctx, mcsServiceAccount); err != nil && errors.IsNotFound(err) {
+			deleteErrors = append(deleteErrors, fmt.Errorf("failed to delete machine config server pod: %w", err))
+		}
+		if err := p.Client.Delete(ctx, mcsRoleBinding); err != nil && errors.IsNotFound(err) {
+			deleteErrors = append(deleteErrors, fmt.Errorf("failed to delete machine config server pod: %w", err))
+		}
+		if err := p.Client.Delete(ctx, mcsPod); err != nil && errors.IsNotFound(err) {
+			deleteErrors = append(deleteErrors, fmt.Errorf("failed to delete machine config server pod: %w", err))
+		}
+		err = utilerrors.NewAggregate(deleteErrors)
+	}()
+
+	err = wait.PollImmediateUntil(1*time.Second, func() (bool, error) {
```

Comment:
my preference here would be to watch the pod, use an HTTP health check on the pod and only attempt the fetch once the pod is available

Reply:
I tried that but got back to polling based on this discussion https://github.com/openshift/hypershift/pull/304#discussion_r656491266 to keep the library design boundaries scoped. I also found out that trying to implement this as an idempotent reconciler func gets to convoluted needing to watch additional resources and having to consider multiple exit cases before deleting the mcs pod resulting in possible leaks.